File: C++/PopNextRight2.cc
#include <iostream>
#include <vector>
using namespace std;

struct TreeLinkNode {
    int val;
    TreeLinkNode* left;
    TreeLinkNode* right;
    TreeLinkNode* next;
    TreeLinkNode(int x): val(x), left(NULL), right(NULL), next(NULL){}
};

//The iterative on the book is more elegant!
void connect(TreeLinkNode *root) {
    TreeLinkNode* preLevel = root;
    while (preLevel) {
        TreeLinkNode* curLevelPre = NULL;
        TreeLinkNode* curLevelStart = NULL;
        while(preLevel) {
            if (preLevel->left || preLevel->right) {
                if (preLevel->right) {
                    if (preLevel->left) {
                        preLevel->left->next = preLevel->right;
                        if (curLevelPre)
                            curLevelPre->next = preLevel->left;
                    } else if (curLevelPre)
                        curLevelPre->next = preLevel->right;
                    curLevelPre = preLevel->right;
                } else {
                    if (curLevelPre)
                        curLevelPre->next = preLevel->left;
                    curLevelPre = preLevel->left;
                }
                if (!curLevelStart)
                    curLevelStart = preLevel->left ? preLevel->left : preLevel->right;
            }
            preLevel = preLevel->next;
        }
        preLevel = curLevelStart;
    }
}

int main(int argc, char** argv)
{
    TreeLinkNode* test = new TreeLinkNode(1);
    test->left = new TreeLinkNode(2);
    test->right = new TreeLinkNode(3);
    test->left->left = new TreeLinkNode(4);
    test->left->right = new TreeLinkNode(5);
    test->right->right = new TreeLinkNode(7);
    connect(test);
    return 0;
}

File: Java/PopNextRight2.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class PopNextRight2 {
    public void connect(TreeLinkNode root) {
        for(TreeLinkNode cur = root; cur != null;) {
            TreeLinkNode pre = null;
            TreeLinkNode nextCur = null;
            for(; cur != null; cur = cur.next) {
                TreeLinkNode firstNonNullNode = null;
                if (cur.left != null) {
                    firstNonNullNode = cur.left;
                    cur.left.next = cur.right;
                } else if (cur.right != null) {
                    firstNonNullNode = cur.right;
                }
                if (pre != null) {
                    pre.next = firstNonNullNode;
                }
                if (nextCur == null) {
                    nextCur = firstNonNullNode;
                }
                pre = (cur.right != null) ? cur.right : ((cur.left != null) ? cur.left : pre);
            }
            cur = nextCur;
        }
    }

    //Second try
    public void connect(TreeLinkNode root) {
        for (TreeLinkNode curNode = root; curNode != null;) {
            TreeLinkNode nextStartNode = null;
            TreeLinkNode preNode = null;
            for (;curNode != null; curNode = curNode.next) {
                if (curNode.left != null && curNode.right != null) {
                    curNode.left.next = curNode.right;
                } 
                if (preNode != null && (curNode.left != null || curNode.right != null)) {
                    preNode.next = (curNode.left != null) ? curNode.left : curNode.right;
                }
                if (curNode.left != null || curNode.right != null) {
                    preNode = (curNode.right != null) ? curNode.right : curNode.left;
                }
                if (nextStartNode == null) {
                    nextStartNode = (curNode.left != null) ? curNode.left : curNode.right;
                }
            }
            curNode = nextStartNode;
        }
    }
}